src/test/java/org/mabrarov/exceptionsafety/FactoryMethodTest.java

```diff
@@ -41,9 +43,33 @@
 
   private OutputStream createConfiguredResource()
       throws IOException, TestResourceConfigurationException {
-    final OutputStream resource = createResource();
-    configureResource(resource);
-    return resource;
+    OutputStream resource = createResource();
+    Throwable throwable = null;
+    Method method = new Object() {
+    }.getClass().getEnclosingMethod();
+    try {
+      configureResource(resource);
+      return resource;
+    } catch (final Throwable e) {
+      throwable = e;
+      for (Class exceptionClass : method.getExceptionTypes()) {
+        if (exceptionClass.isAssignableFrom(e.getClass())) {
+          throw e;
+        }
+      }
+      if (e instanceof RuntimeException || e instanceof Error) {
```

Comment:
This check is reasonable if there is a need to cast `e` to particular type of exception in order to be able to throw it and stay within exception signature of the method. Like:

```java
if (e instanceof RuntimeException) {
  throw (RuntimeException) e;
}
```

You don't cast `e` and code compiles, so, probably this code:

```java
for (Class exceptionClass : method.getExceptionTypes()) {
  if (exceptionClass.isAssignableFrom(e.getClass())) {
    throw e;
  }
}
if (e instanceof RuntimeException || e instanceof Error) {
  throw e;
}
throw new AssertionError("Should never come here", e);
```

can be simplified till:

```java
throw e;
```